### PR TITLE
Basic execution plugin module

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

This adds a basic execution plugin model, very simple to be honest. Most of the work has been done by @sthaha. In a gist this is how it works

- If the specified subcommand doesn't exists, we check if a command
  with the subcommand prefixed by `tkn-` exists in the
  `$HOME/.config/tkn/pulgins` (for `foo` we look for `tkn-foo`).
- The default folder (`$HOME/.config/tkn/plugins`) can be overriden
  with `TKN_PLUGINS_DIR`.
- If the binary doesn't exists, we proceed as usual
- If the binary exists, we execute it "à-la" `exec …`, aka we give full hand to the binary.

```bash
$ ls ~/.config/tkn/plugins/tkn-*
/home/vincent/.config/tkn/plugins/tkn-bar  /home/vincent/.config/tkn/plugins/tkn-foo
$ tkn foo is bar --hello
args is bar --hello
fooo
bar
$ tkn bar bar
/usr/bin/env: ‘invalidcommand’: No such file or directory
```

What may need to be enhanced :
- <del>This needs quick & dirty tests</del>
- This doesn't pass *any* global flags (value or as arguments — `tkn` doesn't have any it seems) — any "client handling" is on the plugin binary to handle
- This doesn't take into account `help` — `tkn foo --help` would call `tkn-foo --help`, *but* `tkn help` won't show `foo` as subcommand. This can probably be taken care of in follow-ups ?
- <del>This only look into `PATH`. Maybe we want to have some "standard" </del>

Carrying #1136.
Fixes: #1039
Signed-off-by: Sunil Thaha <sthaha@redhat.com>

/cc @tektoncd/cli-maintainers 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Run the code checkers with `make check`
- [x] Regenerate the manpages, docs and go formatting with `make generated`
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```release-note
`tkn` now support a basic plugin system. Any command prefixed by `tkn-` in the `PATH` can be used as a `tkn` subcommands (e.g. `tkn-foo` can be used as `tkn foo`).
```

### CODE REVIEW

File: cmd/tkn/main.go
Comment (sthaha) [inline cmd/tkn/main.go:43]: How about `fmt.Fprintf(os.Stderr, ...`

File: cmd/tkn/main.go
Comment (sthaha) [inline cmd/tkn/main.go:44]: why 254? should it be 127? `Command not found or not in path` ? doesn't `128+<exit-code>` have a different meaning? 🤔

File: cmd/tkn/main.go
Comment (sthaha) [inline cmd/tkn/main.go:34]: we should write a  test for this shouldn't we?  ' at least a shell test ?

File: cmd/tkn/main.go
Comment (sthaha) [inline cmd/tkn/main.go:50]: won't this crash if `cmd == nil` ?

File: cmd/tkn/main.go
Comment (vdemeester) [reply cmd/tkn/main.go:34]: @sthaha yeah it's on the desc of the PR :grin:

File: cmd/tkn/main.go
Comment (vdemeester) [reply cmd/tkn/main.go:44]: Random number :sweat_smile: I'll update to 127 :stuck_out_tongue_closed_eyes:

File: cmd/tkn/main.go
Comment (vdemeester) [reply cmd/tkn/main.go:50]: @sthaha right, it will, but `cmd` is never `nil`. But this is an *artefact* of a refactoring I tried :sweat_smile: it should be `tkn.Execute()`, I'll update :sweat_smile:

File: cmd/tkn/main.go
Comment (SM43) [inline cmd/tkn/main.go:35]: If I do only `tkn` with no args then this will break here. index out of range.
Should we handle this? If there is no argument then skip this.

File: cmd/tkn/main.go
Comment (SM43) [inline cmd/tkn/main.go:33]: Why `tkn.Find(args)` doesn't return an error when an invalid command is passed? 
I tried with another CLI `Find` returns err when the command is invalid, it happens only with `tkn`. :smile:

Comment (vdemeester) [reply cmd/tkn/main.go:33]: @SM43 my guess is the `suggest` part but I am not sure :sweat:

**Comment by vdemeester** [comment]
This should be ready to be reviewed. Fixed some problems *and* added a small e2e test :upside_down_face:

File: cmd/tkn/main.go
Comment (vdemeester) [reply cmd/tkn/main.go:36]: fixed it by checking `len(args)` :wink:

**Comment by vdemeester** [comment]
```
TestPipelineRunLogE2E: pipelinerun_test.go:55: could not get pipelinerun: conversion webhook for tekton.dev/v1beta1, Kind=PipelineRun failed: conversion failed to version v1alpha1 for type [kind=PipelineRun group=tekton.dev version=v1beta1] -  the specified field/section is not available in v1alpha1
```
:thinking:

**Review by sthaha — COMMENTED** [review]
looks good to me except of the the unrelated CI failure.

**Review by sthaha — COMMENTED** [review]
/lgtm

**Review by danielhelfand — COMMENTED** [review]
A few thoughts:
* I am wondering if the default should not just look at path but something like `~/.config/tkn` to help users separate out `tkn` plugins from what else is on their PATH
* We should follow up with documentation on plugin usage
* We should consider adding management features around plugins (e.g. `tkn plugin ls` to show available plugins)

It's a good starting point though and am excited to have it.

/lgtm

**Comment by vdemeester** [comment]
> A few thoughts:
> 
>     * I am wondering if the default should not just look at path but something like `~/.config/tkn` to help users separate out `tkn` plugins from what else is on their PATH

Yes, that's a good point, that would be worth deciding before merging this :upside_down_face: (not a fan of `~/.config/tkn` to store binaries but other than that, might make sense to look into a specific path instead of `$PATH`).

>     * We should follow up with documentation on plugin usage

:+1: 

>     * We should consider adding management features around plugins (e.g. `tkn plugin ls` to show available plugins)

Yes, this is a whole other discussion and should be done in a follow-up if this *experimentation* is sucessful :stuck_out_tongue:.

**Comment by danielhelfand** [comment]
> Yes, that's a good point, that would be worth deciding before merging this 🙃 (not a fan of ~/.config/tkn to store binaries but other than that, might make sense to look into a specific path instead of $PATH).

Something that mirrors `kn`'s plugin model location would make sense (`~/.config/tkn/plugins`). I would be fine with moving forward with starting with `$PATH` and supporting `~/.config/tkn` or something of the like later on though. I see no reason why we can't make it work for both.

**Comment by vdemeester** [comment]
> Something that mirrors `kn`'s plugin model location would make sense (`~/.config/tkn/plugins`). I would be fine with moving forward with starting with `$PATH` and supporting `~/.config/tkn` or something of the like later on though. I see no reason why we can't make it work for both.

Right :sweat_smile: I'd rather start with supporting only `~/.config/tkn/plugins` at first and add support to `$PATH` if there is a need. I'll update :wink:

**Comment by vdemeester** [comment]
Updated, `$HOME/.config/tkn/plugins` is the default folder, and you can override it with `TKN_PLUGINS_DIR` environment variable.

**Comment by vdemeester** [comment]
This is really simple, it doesn't support multiple folder or using the `PATH`. That can be done later on.

**Comment by danielhelfand** [comment]
Rebasing around master with #1249 should resolve EventListener test failures

File: cmd/tkn/main.go
Comment (danielhelfand) [inline cmd/tkn/main.go:31]:
```suggestion
	pluginDir    = "~/.config/tkn/plugins"
```

**Review by danielhelfand — COMMENTED** [review]
Looks good otherwise, just a suggestion for the default plugin path.

File: cmd/tkn/main.go
Comment (vdemeester) [reply cmd/tkn/main.go:31]: Oh right, I wanted to use `homedir` to expand this :sweat_smile:

Comment (vdemeester) [reply cmd/tkn/main.go:32]: Updated, using `homedir.Expand` now :wink:

**Comment by vdemeester** [comment]
https://github.com/tektoncd/cli/pull/1252 fixes the cross error

**Review by danielhelfand — APPROVED** [review]

**Comment by chmouel** [comment]
Is `~/.config/tkn` the right place from a LSB distro/point of view ? it should be ~/.local/share isn't it ? But most people would be confused and would not care about FHS or LSB i guess, can we have `~/.config/tekton/plugins` tho ? in case we end up with other stuff we want that is tekton related but not tkn releated.

krew has it in `kube/plugins`

**Comment by vdemeester** [comment]
> Is `~/.config/tkn` the right place from a LSB distro/point of view ? it should be ~/.local/share isn't it ? But most people would be confused and would not care about FHS or LSB i guess, can we have `~/.config/tekton/plugins` tho ? in case we end up with other stuff we want that is tekton related but not tkn releated.
> 
> krew has it in `kube/plugins`

krew has `.krew/bin` doesn't it ? I choosed `tkn` instead of `tekton`, mainly because the tool is called `tkn` :sweat_smile:. If there is another cli tool for tekton apis named `foo` in the future, what make `tkn` more inclined to use `tekton` instead of `foo` ?

I don't have strong opinion on this though, I just want us to go with a given setup and see how it goes and how it is useful.

**Comment by vdemeester** [comment]
@tektoncd/cli-maintainers are we good to merge given the result of the [poll](https://doodle.com/poll/n5iit5dayrsyxcyy) ?

**Review by piyush-garg — APPROVED** [review]

**Comment by chmouel** [comment]
/lgtm

Looking Good! 🤙🏽

**Pull request merged**
